lib/membership_pair_widget.dart

```diff
@@ -0,0 +1,175 @@
+    final onIceCandidate = (RTCIceCandidate iceCandidate) {
+      membershipPairChannelState.value.createWebRtcIceCandidate(
+        sdp: iceCandidate.candidate,
+        sdpMid: iceCandidate.sdpMid,
+        sdpMlineIndex: iceCandidate.sdpMlineIndex
+      );
+    };
+    final onAddStream = (MediaStream stream) => remoteStreamState.value = stream;
+
+    useEffect(() {
+        Future.wait([
+          localStreamFuture,
+          createPeerConnection(PEER_CONFIG, {})
+        ]).then((resolvedFutures) {
+          final MediaStream localStream = resolvedFutures[0];
+          final RTCPeerConnection peerConnection = resolvedFutures[1];
+
+          peerConnection.addStream(localStream);
+          peerConnection.onSignalingState = onSignalingState;
+          peerConnection.onIceCandidate = onIceCandidate;
+          peerConnection.onAddStream = onAddStream;
+
+          peerConnectionState.value = peerConnection;
+        });
+
+        return () => peerConnectionState.value?.dispose();
+    }, []);
+
+    useEffect(() {
+      if (peerConnectionState.value == null) {
+        return;
+      }
+
+      () async {
+        final peerConnection = peerConnectionState.value;
+        if (membershipPairEntry.older && peerConnectionSignalingStateState.value == RTCSignalingState.RTCSignalingStateStable) {
+          final offer = await peerConnection.createOffer(SESSION_CONSTRAINTS);
+          await peerConnection.setLocalDescription(RTCSessionDescription(offer.sdp, offer.type));
+          membershipPairChannelState.value.createWebRtcOffer(sdp: offer.sdp);
+        } else if (membershipPairState.value.webRtcOffer != null && addedWebRtcOffer == null) {
+          final webRtcOffer = membershipPairState.value.webRtcOffer;
+          await peerConnection.setRemoteDescription(RTCSessionDescription(webRtcOffer.sdp, 'offer'));
+          addedWebRtcOffer.value = webRtcOffer;
+
+          final answer = await peerConnection.createAnswer(SESSION_CONSTRAINTS);
+          await peerConnection.setLocalDescription(RTCSessionDescription(answer.sdp, answer.type));
+          membershipPairChannelState.value.createWebRtcAnswer(sdp: answer.sdp);
+        } else if (membershipPairState.value.webRtcAnswer != null && addedWebRtcAnswer == null) {
+          final webRtcAnswer = membershipPairState.value.webRtcAnswer;
+          await peerConnection.setRemoteDescription(RTCSessionDescription(webRtcAnswer.sdp, 'answer'));
+        }
+
+        final canAddIceCandidates = [
+          RTCSignalingState.RTCSignalingStateHaveLocalPrAnswer,
+          RTCSignalingState.RTCSignalingStateHaveRemotePrAnswer
+        ].contains(peerConnectionSignalingStateState.value);
+        if (!canAddIceCandidates) {
+          return;
+        }
+
+        final webRtcIceCandidatesToAdd = Set.from(membershipPairState.value.webRtcIceCandidates).difference(addedWebRtcIceCandidatesState.value);
+        await Future.wait(webRtcIceCandidatesToAdd.map((webRtcIceCandidate) async {
```

Comment:
is this too cute?

felt weird to await each of these one by one, but this might be hard to understand